README
*This qvm is no longer in active developement. If you have requests, feel free to do so here or email me at my youtube channel
youtube.com/user/zdrytchx

CoW Tremulous QVM v2.0 [incomplete] merge with GPP-1.1/Lolards v0.7 mod

Credits for bots:
The original works of vcxzet (lamebot3) were used a guide to create TremBot.
The works of Amine Haddad (TremBot) and Sex (PBot) where used/modified to create COW Bot

From CoW bot code by Fuma (not up-to-date):
{
 Current Features (Over Lakitu QVM 5.5):
 Advanced bots with route finding capabilities and waypoint support. Entirely customisable!
 !loadlayout allows loading of a new layout without a restart
 !loadrotation allows the changing of rotations on demand
 !strip and associated commands allow admins to control a player's access to weapons/armor/classes. 
 
 Bot Specific Features:
 !botcfg allows loading/saving of bot configuration files
 !botset allows the changing of bot abilities
 !botcmd changes the bots behavior
 !bot adds bots with a specified skill level
 !drawnodes shows the waypoints
 /node allows the making of waypoints for the bots  

 Bots can:
 Go to a medistat to heal
 Go to an armoury to buy weapons and armor
 Evolve
 Use secondary abilities like lighting, pounce, snipe, charge
 Wallcrawl
 Use waypoints to navigate the map and go to a location
 Dodge.
 Will not keep human players in a long spawn queue
 Anti-Hax-type aiming using recoil system depending on skill level
}
Great thanks to Fuma for helping me with setting up a github account and helping me with the code
See his source code @ https://github.com/cmf028/CoW-Tremulous
FYI his version number needs to increment to 3.0. Don't get the wrong idea that i've upgraded to the new one.

Also, thanks /dev/humancontroller for debugging and Volt for helping me start modding in the early days with his tutorial pack.

==Bot configuration edits by me==
-Goon pounces TO human, not ABOVE human.
-Bot Lcannon charges 2.5s rather than original 2
-Bots now don't sit in the same spot when killing a structure
-Maras aren't so bloody overpowered
-

GPP-1.1 Gameplay "Features":
-Lil' Gwangie has his cutie tusks available in combat
-Smooth Regeneration
-Regen speeds depend upon creeps and basis
-Rant can "lawn mow" but has beenn nerfed by lots, and does 20% damage to buildables only
-Pounce whoring possible, but has been balanced with chomps
-Re-gassing as a basilisk resets the victim's disorientation timer rather than doing nothing and waiting for it to finish before becomming effective
-Gas goes through armour
-Gas deals a mere 1 damage per second (code-wise it deals 0 damage though, dunno whats wrong here)
-Lcannon is a 1.1-Unvanquished mix (Dynamic luci ball speed from Lolards, 2.5 second primary charge time, gpp-like secondary or finish-off combos)
-Most buildables take gpp values
-Turrets are a mix - Crappy at close range againts a fast alien, but great in damage dealing and fires inaccurately (misses dretches a lot long range, read Lolards features)
-DCC simply boosts turret aiming speeds and distribute defences (like 1.1 ) but aren't required for tesla generators
-No dcc required, but human buildables regenerate slowly (~2hp/s for most) rather than gpp's 4hp/s/dcc_in_range
-Painsaw range 60, not 40, not 80 so it is able to reach under feet without being overpowered, uses 1.1 damage/s
-GPP Pulse rifle projectile speed, 1.1 clip system
-Unvanquished-like stamina system, but still slows down below 50% to reduce the benifit of strafe jumping
-Battlesuits regen more stamina, Light armour takes more stamina so bsuits are more useful
-Mara zap more useful againts buildables
-Marazap chain count 3 -> 5 (may be overpowered)
-Zap sustains even after victim's death unless a new victim is found or out of range
-Zap sustains after one's death (Try getting a kill while you're dead!)
-Dretche bites 48 -> 36
-Goon barb exploziens (75% damage only because it's overpowered)
-Increased barb reload timer to 15 seconds (gpp = 30, 1.1 = 10)
-GPP shotgun (11 pellets, each doing 5 damage, where 1.1 has 14 pellets, each 4 dmg) to help againts dretches
-Bunny Hopping up stairs is now a tonne load easier
-
-Other shit

"Imported" Gameplay from Lolards/Misc:
-Modified Turret aiming code - they aim slower when firing, and aim even slower when they're aiming directly at you
-Humans regenerate slowly
-Blaster/Flamer jumping/climbing is possible now (blaster splash isnt supposed to deal damage, but its damage is a mere 1, dunno what happened here, look at basi gas above)
-Dynamic luci ball speed (applied before unvanquished did theirs actually)
-Smoother flamer firing speed (10 damage per missile instead of 20, fires twice as fast, twice the ammo)
-Chaingun has a more inaccurate secondary which splits the bullets into three, but fires a tad slower
-Chaingun has an extra clip with a painfully long reload
-Lasgun isn't 100% accurate, although spread is little (anti camping reasons, it's the best survival weapon afterall)
-Non-crappy JETPACK
-Extreme High falls now can insta-kill all armoured humans
-Mara zap has a "sustain" range seperate from the "initialising" range
-Mara zap does not lose as much damage per victim added to the chain
-Zap initialising range 300 -> 200
-Zap "Sustaining" range 300 -> 400
-Other shit

Potential Side Effects/bugs from features:
-sv_fps MUST be >= 40 else aliens will have a hard time regenerating
-Some clients on rare occasions can be "bot infected" (not literally, just a term where his client takes over a bot's and the bot tries to control him when he's lagging) although this bug has not been seen for over half a year
-With alien's smooth regeneration system, the server MAY take more CPU (not confirmed)



ChangeLog
//lakitu 7 only. None are related to my qvm.
1.1.0
-----

    Feature/balance changes
    -----------------------
    * Quick fullscreen changes
    * Fancy autocompletion
    * Client now sleeps when inactive
    * Persistent console history
    * Increase boost time to 30 seconds from 20 seconds
    * Lcannon now has a minimum charge value
    * Flamer damage down to 20 from 31
    * Zap repeat rate down to 1500ms from 2000ms
    * Knockback from pouncing now 3 times as strong
    * Demo state now displayed from cgame
    * AVI video capture
    * cl_autoRecordDemo
    * Master server, based on dpmaster
    * Changed defaults for r_picmip and r_textureMode
    * Chat sounds differ depending on team
    * New collision type, for improved light flares
    * New pain blend effect
    * Reduced Dragoon pounce range to 64 from 96
    * Dragoon pounce now incurs 400ms wait before other weapons may be used
    * Trapper cost down to 8 from 10
    * Hovel is now free, but only one can be built
    * Teslagen damage up to 9 from 7
    * Non locational damage does not apply locational armour anymore; instead it
      averages the armour regions together //which doesn't work well :P
    * Added target_alien_win and target_human_win
    * Added worldspawn keys to disable specific game elements
    * 3D particles
    * Brass ejections now done via particle system
    * static_tranform particle move type
    * thirdPersonOnly particle and trail system property
    * Abstract attachment system
    * Scriptable trails system

    Bug fixes/development issues
    ----------------------------
    * Fix to a server memory leak
    * Fix to a server 100% CPU bug
    * Delete key on *nix fixed
    * Fixed lcannon charge storage exploit
    * Fixed psaw/cgun inappropriate effects bug
    * Maps on create server menu now sorted by name
    * Fixed death by poison MOD s/antitox/medkit/
    * Fixed Marauders momentarily disappearing when wall jumping
    * Fixed a potential crash bug involving the use of generic1
    * Fixed being able to build multiple coincident repeaters if there is no reactor
    * Fixed incorrect message when invoking "buy ammo" with an energy weapon and no
      reactor present
    * Fixed invoking "reload" during a weapon reload causing an unnecessary reload
    * Fixed aliens having the wrong blood colour when shot with a las gun
    * Fixed hovel causing invisible builders

1.0.2
-----

    Feature/balance changes
    -----------------------
    * Extended STATS report to include the times when stages change
    * Increased Tesla damage frequency
    * Marginally increased Tesla damage
    * Teslas now produce exaggerated knockback effect
    * Hives now buildable on walls
    * Boosters no longer restore Dragoon spitballs
    * Dragoon spitballs now affected by gravity
    * Dragoon pounce attack damage reduced to 100 from 160
    * Removed jump pad code (and associated media loading)
    * Removed loading of some Q3 cruft
    * Corpses timeout in 20 seconds instead of 60

    Bug fixes/development issues
    ----------------------------
    * Fixed a silly bug with using the reactor/repeater
    * Spilled events attached to temporary entities are now reattached
      to their original entities, fixing the missing flame bug
    * Fixed "suicide god" bug
    * Fixed bug where zap ignores armour
    * Fixed missing particle systems when follow-spectating
    * Fixed potential bug involving dodgy pointer arithmetic in
    CG_LoadClientInfo
    * Cleaned up logic in CG_AddPlayerWeapon
    * MASK_SHOT traces no longer collide with corpses
    * Improved robustness of spawn validation, fixing the bug on transit
    * A crapload of whitespace fixes

1.0.1
-----

    Feature/balance changes
    -----------------------
    * cg_debugParticles >= 1 now prints to the console when particle
    systems are parsed and registered
    * Reverted walk/run animation switching back to cmd.buttons instead
    of basing it on speed
    * Added option for changing whether or not the wallwalk control is a
    toggle or not
    * Weapon now drops momentarily when reloading
    * The stage kill counters are now incremented for structure kills if
    players did more then 50% of the total damage
    * Removed the mp3 decoder from the source
    * Marauder lightning now requires aim, does damage over time and
    chains to other entities
    * Implemented the Medkit -- a means for a human to restore health
    and cure poison in the field
    * "Disable Build Warnings" replaced with "Disable Warning
    Dialogs" and improved
    * Sped spectator move speed up
    * Implemented "step down" physics for all characters; no more
    jumping down stairs
    * Increased frequency with which the Acid Tube deals damage
    * Tyrant can no longer charge up forever and must pass a specific
    minimum charge level
    * Implemented command queueing for commands sent to clients in order
    to prevent overflows even sv_floodProtect is off, but not by
    dropping commands
    * Added LOS check to creep slowing
    * Overmind now only complains if there are 0 spawns
    * Spawns can no longer be built when there is no Overmind/Reactor
    * The spawn closest to the point of death is chosen preferably if
    available
    * Evolving no longer restores all health
    * "give weapons" and "give ammo" cheats removed
    * When selling the battery pack, max ammo is given
    * Added stage information to the end of game stats
    * Reduced Dragoon spitball damage from 120 to 110
    * Reduced Tyrant claw damage from 120 to 100
    * Reduced Tyrant charge damage from 160 to 110
    * Increased Barricade regeneration rate from 12 to 14
    * Increased Overmind health from 500 to 750
    * Decreased Overmind regeneration rate from 10 to 6
    * Doubled Blaster speed from 700 to 1400
    * Reduced Painsaw damage from 18 to 15
    * Reduced Painsaw range from 48.0 to 40.0
    * Reduced Grenade price from 300 to 200

src/game/g_bot.h
/*
===========================================================================
Copyright (C) 1999-2005 Id Software, Inc.
Copyright (C) 2000-2006 Tim Angus

This file is part of Tremulous.

Tremulous is free software; you can redistribute it
and/or modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the License,
or (at your option) any later version.

Tremulous is distributed in the hope that it will be
useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Tremulous; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
===========================================================================
*/
//g_bot.h - declarations of bot specific functions
//#include "g_local.h"

int G_BotBuyUpgrade ( gentity_t *ent, int upgrade );
void G_BotThink(gentity_t *self);
void G_BotModusManager(gentity_t *self);
void G_BotAttack(gentity_t *self, usercmd_t *botCmdBuffer);
void G_BotRepair(gentity_t *self, usercmd_t *botCmdBuffer);
void G_BotHeal(gentity_t *self, usercmd_t *botCmdBuffer);
void G_BotRoam(gentity_t *self, usercmd_t *botCmdBuffer);
void G_BotBuy ( gentity_t *self, usercmd_t *botCmdBuffer);
void G_BotGoto(gentity_t *self, botTarget_t target, usercmd_t *botCmdBuffer);
void G_BotMoveDirectlyToGoal( gentity_t *self, usercmd_t *botCmdBuffer );
void G_BotReactToEnemy(gentity_t *self, usercmd_t *botCmdBuffer);
void G_BotDodge(gentity_t *self, usercmd_t *botCmdBuffer);
int botGetAimEntityNumber(gentity_t *self);
qboolean botTargetInAttackRange(gentity_t *self, botTarget_t target);
void botFireWeapon(gentity_t *self, usercmd_t *botCmdBuffer);
int getTargetEntityNumber(botTarget_t target);
void getTargetPos(botTarget_t target, vec3_t *rVec);
int getTargetTeam( botTarget_t target);
int getTargetType( botTarget_t target);
qboolean targetIsEntity( botTarget_t target);
void setGoalEntity(gentity_t *self, gentity_t *goal );
void setGoalCoordinate(gentity_t *self, vec3_t goal );
void setTargetEntity(botTarget_t *target, gentity_t *goal );
void setTargetCoordinate(botTarget_t *target, vec3_t goal );

int G_BotEvolveToClass( gentity_t *ent, char *classname, usercmd_t *botCmdBuffer);
void G_BotEvolve ( gentity_t *self , usercmd_t *botCmdBuffer);
void botGetAimLocation( botTarget_t target, vec3_t *aimLocation);
void botSlowAim( gentity_t *self, vec3_t target, float slow, vec3_t *rVec);
void botShakeAim( gentity_t *self, vec3_t *rVec );
void botAimAtLocation( gentity_t *self, vec3_t target , usercmd_t *rAngles);
int botFindDamagedFriendlyStructure( gentity_t *self );
int botFindClosestEnemy( gentity_t *self, qboolean includeTeam );
int botFindBuilding(gentity_t *self, int buildingType, int range);

int findClosestNode( botTarget_t target );
void findNewNode(gentity_t *self, usercmd_t *botCmdBuffer);
void findNextNode(gentity_t *self);
void doLastNodeAction(gentity_t *self, usercmd_t *botCmdBuffer);
int distanceToTargetNode(gentity_t *self);
void findRouteToTarget( gentity_t *self, botTarget_t target );
void setNewRoute(gentity_t *self);

qboolean botTargetInRange( gentity_t *self, botTarget_t target, int mask );
qboolean G_BotCheckForSpawningPlayers( gentity_t *self );
qboolean botWeaponHasLowAmmo(gentity_t *self);
void setSkill(gentity_t *self, int skill);
int getStrafeDirection(gentity_t *self);
qboolean botPathIsBlocked(gentity_t *self);
qboolean botShouldJump(gentity_t *self);
qboolean botNeedsItem(gentity_t *self);
qboolean botCanShop(gentity_t *self);
qboolean botStructureIsDamaged(int team);
int G_BotBuyWeapon(gentity_t *ent, int weapon);
//configureable constants
//For a reference of how far a number represents, take a look at tremulous.h

//how far the bots can be from a medistat to walk to it to heal, if they are too far away, they will not go to it
#define BOT_MEDI_RANGE 1000.0f //1500 -> 1000

//how far the bots can be from an armoury to walk to it to buy stuff, if they are too far, they will not go to the arm
#define BOT_ARM_RANGE 1500.0f

//when closer to the enemy than this range, the human bots will backup
#define BOT_BACKUP_RANGE 400.0f //300 -> 400

//How long in milliseconds the bots will chase an enemy if he goes out of their sight (humans) or radar (aliens)
#define BOT_ENEMY_CHASETIME 15000 //5000

//How long in milliseconds the bots will chase a friend if he goes out of their sight (humans) or radar (aliens)
#define BOT_FRIEND_CHASETIME 15000 //5000

//How often in milliseconds, we will search for a new (closer) enemy this needs to be kept <= 10000 for now
#define BOT_ENEMYSEARCH_INTERVAL 3000 //500 //1500 -> 3000, they keep changing too often

//at what hp do we use medkit?
#define BOT_USEMEDKIT_HP 48 //50

//when human bots reach this ammo percentage left or less(and no enemy), they will head back to the base to refuel ammo when in range of arm as defined by BOT_ARM_RANGE
#define BOT_LOW_AMMO 0.30f //0.50f

//when human bots reach this health or below (and no medkit/enemy) they will head back to the base to heal when in range of medi as defined by BOT_MEDI_RANGE
#define BOT_LOW_HP 30 //100 - now 60 because of slow autoregen + vamp //removed vamp

//TODO: implement the rest of these, currently they do nothing :)

//when the bots get closer than this distance to an enemy Egg/Node, they will head toward it (killing stuff in the way if need be) instead of roaming randomly
//Line of sight to the egg/node does not matter
#define ROUTETO_SPAWN_RANGE 2000.0f 

//If the "The base is under attack" message appears, the bots will head back to the base to defend it, if they are at least this close to the base (and no enemy in range).
#define H_ROUTETO_BASE_RANGE 5000.0f

//When the "The Overmind is under Attack!" message appears, the alien bots will head back to the base to defend it, if they are at least this close to it.
#define A_ROUTETO_BASE_RANGE 5000.0f

//ZdrytchX: New variables here by me.
#define BUILDABE_REPAIR_HEALTH 0.5 //How much minimum health a buildable must have before it is to be healed
